core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,374 @@
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0L) {
+                close(new IllegalArgumentException(
+                        "Rule §3.9 violated: non-positive subscription requests are forbidden."));
+                return;
+            }
+
+            if (closed) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                request0(n);
+            } else {
+                executor.execute(() -> request0(n));
+            }
+        }
+
+        private void request0(long n) {
+            final long oldRequested = requested;
+            if (oldRequested == Long.MAX_VALUE) {
+                return;
+            }
+            if (n == Long.MAX_VALUE) {
+                requested = Long.MAX_VALUE;
+            } else {
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // InputStreamSubscription is reading input stream.
+                // New requests will be handled by 'publishDownstream(HttpData)'.
+                return;
+            }
+
+            readBytes();
+        }
+
+        private void readBytes() {
+            if (requested <= 0) {
+                return;
+            }
+
+            blockingTaskExecutor.execute(() -> {
+                if (position >= end) {
+                    close(null);
+                    return;
+                }
+
+                if (position < offset) {
+                    final long skip = offset - position;
+                    final long actualSkipped;
+                    try {
+                        actualSkipped = inputStream.skip(skip);
+                    } catch (IOException e) {
+                        close(e);
+                        return;
+                    }
+
+                    if (actualSkipped < skip) {
+                        close(null);
+                        return;
+                    }
+
+                    position += skip;
+                }
+
+                final int bufferSize = Math.min(this.bufferSize, Ints.saturatedCast(end - position));
+                final int len;
+                final byte[] readBytes = new byte[bufferSize];
+                try {
+                    len = inputStream.read(readBytes);
+                } catch (IOException e) {
+                    close(e);
+                    return;
+                }
+
+                if (len == -1) {
+                    close(null);
+                    return;
+                }
+
+                final HttpData data = HttpData.wrap(readBytes, 0, len);
+                position += len;
+
+                executor.execute(() -> publishDownstream(data));
+            });
+        }
+
+        private void publishDownstream(HttpData data) {
+            if (closed) {
+                return;
+            }
+
+            downstream.onNext(data);
+            requested--;
```

Comment:
I think it might be better to decrement `requested` just in case `request()` is called again synchronously

```suggestion
            requested--;
            downstream.onNext(data);
```

Reply:
✅ updated